controller/array_action/array_mediator_svc.py

```diff
@@ -512,10 +506,41 @@ def copy_to_existing_volume(self, volume_id, source_id, source_capacity_in_bytes
         target_volume_name = self._get_volume_name_by_wwn(volume_id)
         self._copy_to_target_volume(target_volume_name, source_name)
 
-    def create_volume(self, name, size_in_bytes, space_efficiency, pool, io_group, volume_group):
-        self._create_cli_volume(name, size_in_bytes, space_efficiency, pool, io_group, volume_group)
+    def _create_volume_in_volume_group(self, name, pool, io_group, source_id):
+        self._mkvolumegroup(name, pool, io_group, source_id)
+
+    def _fix_creation_side_effects(self, name, volume_id, volume_group):
+        try:
+            self._change_volume_group(volume_id, volume_group)
+            self._rmvolumegroup(name)
+            self._rename_volume(volume_id, name)
+        except (svc_errors.CommandExecutionError, CLIFailureError, array_errors.VolumeAlreadyExists) as ex:
+            self._rollback_create_volume_from_snapshot(volume_id, name)
+            raise ex
+
+    def _create_cli_volume_from_snapshot(self, name, pool, io_group, volume_group, source_id):
+        logger.info("creating volume from snapshot")
+        self._create_volume_in_volume_group(name, pool, io_group, source_id)
+        volume_id = self._get_volume_id_from_volume_group(name)
+        self._fix_creation_side_effects(name, volume_id, volume_group)
+
+    def _create_cli_volume_from_source(self, name, pool, io_group, volume_group, source_ids, source_type):
+        if source_type == controller_config.SNAPSHOT_TYPE_NAME:
+            self._create_cli_volume_from_snapshot(name, pool, io_group, volume_group, source_ids.internal_id)
+
+    def _is_source_support_addsnapshot(self, object_uid):
```

Comment:
```suggestion
    def _is_source_support_addsnapshot(self, vdisk_uid):
```
same for `_is_source_has_fcmaps`